Add more documentation checks

jngrad opened the pull request:
Closes #2661

Description of changes:
- broken links in Sphinx are now treated as errors
   + mostly incorrect namespaces, typos and incorrect names (e.g. `integer` instead of `int`)
- some Doxygen warnings are now treated as errors
   + warnings caused by known Doxygen bugs are ignored
   + warnings that don't change the html output are ignored
   + only unique warnings are displayed
- developers can use `maintainer/CI/doc_warnings.sh` and `maintainer/CI/dox_warnings.sh` to list Sphinx resp. Doxygen issues

jngrad committed:
Add more documentation checks in CI

Check for broken links in the Sphinx documentation and for common
warnings in the Doxygen documentation.

jngrad committed:
Formatting

jngrad committed:
Fix for container cuda:9.0

jngrad committed:
espresso-ci: post documentation warnings to the PR

Script `maintainer/gh_post_docs_warnings.py` is derived from
script `maintainer/gh_post_style_patch.py`.

jngrad committed:
Filter out false positives

Skip exception classes, skip espressomd.visualization module.

jngrad committed:
Doxygen: fix ScriptInterface warnings

jngrad committed:
Doxygen: fix P3MParameters warnings

jngrad committed:
Doxygen: fix remaining warnings

jngrad committed:
Remove legacy code

jngrad committed:
Sphinx: fix names of built-in types

jngrad committed:
Sphinx: fix broken links

KaiSzuttor commented on maintainer/CI/dox_warnings.py:
```diff
@@ -0,0 +1,97 @@
+        # defined in another group in the .cpp file; this is usually caused by
+        # the "Private functions" and "Exported functions" groups in .hpp files
+        continue
+    if re.search('^documented symbol `\S+\' was not declared or defined\.$', warning):
```
I'm wondering if all this is maintainable... @fweik what do you think?

jngrad committed:
Fix exit code and filter out false positives

Simplify code logic. Skip links to features not compiled in CI and
to private classes/methods (one leading underscore in their name),
which cannot be linked to. Check the Doxygen version.

pkreissl commented on doc/sphinx/magnetostatics.rst:
```diff
@@ -171,11 +171,11 @@ calculations. The feature ``SCAFACOS_DIPOLES`` has to be added to
 :file:`myconfig.hpp` to activate this feature. At the time of this writing (Feb
```
This should probably go: "At the time of this writing (Feb 2018)"

jngrad replied:
done

jngrad commented:
Progress: from 136 Doxygen warnings to 0, from 66 Sphinx broken links to 0. The scripts worked as expected: no container failure, previous espresso-ci bot messages automatically removed.

Next step: merge python into this branch, check the espresso-ci bot posts the new warnings, fix the new warnings, then merge before anyone else can merge.

jngrad committed:
Improve maintainer scripts

Skip source code html pages (_module subfolder in Sphinx), notify user
that some broken links won't be checked in CI, list probable issues
with reSt syntax parsing (without failing CI).

jngrad committed:
Sphinx: fix broken syntax

jngrad commented:
The syntax checker implemented in 68f1a5b correctly identified the broken roles ([logfile](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/110107)). They were corrected in cde9732, and now the trace only shows `Found no broken link requiring fixing` ([logfile](https://gitlab.icp.uni-stuttgart.de/espressomd/espresso/-/jobs/110135)).

We won't make these syntax errors fail CI because they are difficult to spot in the HTML output, and it's not always obvious why they fail (e.g. the `:rtype:`appeared in the HTML output because a Sphinx external plugin incorrectly parsed `Returns` section headings in `cuda_init.pyx`).

jngrad committed:
Merge branch 'python' into warnings-ci

fweik commented:
bors r+
